### PR TITLE
Add FedNova baseline 

### PR DESCRIPTION
**Pull request by ashdtu**
## FedNova

Open PR(in progress) for implementation of FedNova baseline to Flower framework under Summer of Reproducibility challenge.

### CODE REVIEW

**Comment by jafermarq** [comment]
Hi @ashdt, how is it going? Let me know if you'd like to discuss something about your `FedNova` implementation. The _Summer of Reproducibility_ ends in ~10 days.

**Comment by ashdtu** [comment]
Hi @jafermarq I have a working version of the baseline now. that is producing results similar to the one in the paper. However there are certain parts that I wanted to refactor in the code that will be a work in progress in the coming week. Request you to take a look at the current state so that we can have a 1-on-1 soon.

File: baselines/fedNova/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+****Dataset:**** :warning: *_Earlier you listed already the datasets that your baseline uses. Now you should include a breakdown of the details about each of them. Please include information about: how the dataset is partitioned (e.g. LDA with alpha 0.1 as default and all clients have the same number of training examples; or each client gets assigned a different number of samples following a power-law distribution with each client only instances of 2 classes)? if  your dataset is naturally partitioned just state “naturally partitioned”; how many partitions there are (i.e. how many clients)? Please include this an all information relevant about the dataset and its partitioning into a table._*
+
+****Training Hyperparameters:**** :warning: *_Include a table with all the main hyperparameters in your baseline. Please show them with their default value._*
+
+
+## Environment Setup
+
```
Comment (jafermarq) [inline baselines/fedNova/README.md:48]: I know you are still working on this but let me add some content that would be great to have (this is what essentially other baselines are doing) -- add it as a code block:
```
# Set python version
pyenv install 3.10.6
pyenv local 3.10.6

# Tell poetry to use python 3.10
poetry env use 3.10.6

# install the base Poetry environment
poetry install

# activate the environment
poetry shell
```

File: baselines/fedNova/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+:warning: _The Python environment for all baselines should follow these guidelines in the `EXTENDED_README`. Specify the steps to create and activate your environment. If there are any external system-wide requirements, please include instructions for them too. These instructions should be comprehensive enough so anyone can run them (if non standard, describe them step-by-step)._
+
+
+## Running the Experiments
+
```
Comment (jafermarq) [inline baselines/fedNova/README.md:53]: Ideally, baselines are executed like this (after having done `poetry shell`):

```
python -m fednova.main # optionally passing arguments to override the base config
```

File: baselines/fedNova/fedNova/models.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+			local_tau = torch.tensor(self.local_normalizing_vec * self.ratio)
+		local_stats = {"weight": self.ratio, "tau": local_tau.item(), "local_norm": self.local_normalizing_vec}
+
+		return local_stats
+
+	def load_aggregated_grad(self, agg_grad):
+		if len(agg_grad) == 0:
+			i = 0
+			# print("---------------No parameters received, filling it with init params------------")
+			init_params = torch.load("state/init_params_torch.pt")
```
Comment (jafermarq) [inline baselines/fedNova/fedNova/models.py:260]: Can't this be passed from the main instead of having to load it from disk?

**Review by jafermarq — COMMENTED** [review]
Hi @ashdtu,

I have done an initial pass. See some suggestions below. The summary of my comments are:
* (important) please rename your directory (and its subdirectory) from `fedNova` to `fednova`. Then don't forget to reflect that in the `pyproject.toml` (see comment), and update all the imports so they make use of `fednova.<your file>`. I was able to run your code after doing it by doing `python -m fednova.main`.
* Try to have a minimal version of the readme ready.
* If you make plots, please add them to a directory called `_static` and placed in the same directory as the `README.md`.
* Does `state/init_params_torch.pt` need to be part of the repo? (it's a large file)

Please ping me when you are done making the changes. Happy to discuss them further over here or slack.

File: baselines/fednova/fednova/strategy.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            self.global_parameters: List[NDArray] = parameters_to_ndarrays(
+                self.initial_parameters
+            )
+
+        self.exp_config = exp_config
+        self.lr = exp_config.optimizer.lr
+        self.gmf = exp_config.optimizer.gmf
+        self.best_test_acc = 0.0
+
+    def aggregate_fit(
```
Comment (jafermarq) [inline baselines/fednova/fednova/strategy.py:41]: Could you add a couple of comments in this method to point to specific lines in the FedNova algorithm? (just to help people navigate the code more easily)

File: baselines/fednova/fednova/strategy.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            scale = tau_eff / float(res.metrics["local_norm"])
+            scale *= float(res.metrics["weight"])
+            aggregate_parameters.append((params, scale))
+
+        agg_cum_gradient = aggregate(aggregate_parameters)
+        self.update_server_params(agg_cum_gradient)
+
+        return ndarrays_to_parameters(self.global_parameters), {}
+
+    def update_server_params(self, cum_grad: NDArrays):
```
Comment (jafermarq) [inline baselines/fednova/fednova/strategy.py:71]: Could you add a couple of comments in this method to point to specific lines in the FedNova algorithm? (just to help people navigate the code more easily)

File: baselines/fednova/EXTENDED_README.md
Comment (jafermarq) [inline baselines/fednova/EXTENDED_README.md:1]: please remove this file

**Review by jafermarq — COMMENTED** [review]
Hey @ashdtu,

Just a couple of small comments about the README.md and the location where the results are stored.

File: baselines/fednova/fednova/main.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        num_clients=cfg.num_clients,
+        config=fl.server.ServerConfig(num_rounds=cfg.num_rounds),
+        strategy=strategy,
+        client_resources=cfg.client_resources,
+        ray_init_args={"ignore_reinit_error": True, "num_cpus": 8},
+    )
+
+    # 6. Save your results
+    # save_path = HydraConfig.get().runtime.output_dir
+    save_path = cfg.results_dir
```
[the remark this answers is not in the document]
Comment (ashdtu) [reply baselines/fednova/fednova/main.py:117]: I refactored the plot logic in utils.py to accomodate this change and also made it generic to accept any set of experiments for plotting.

File: baselines/fednova/fednova/strategy.py (hunk 5 above)
Comment (ashdtu) [reply baselines/fednova/fednova/strategy.py:86]: Done. Added comments.

File: baselines/fednova/fednova/strategy.py (hunk 4 above)
Comment (ashdtu) [reply baselines/fednova/fednova/strategy.py:44]: Done.

**Review by jafermarq — APPROVED** [review]
Hi @ashdtu,

I did a couple of small changes today. But i'm glad to say `FedNova` looks merge-ready !! Many thanks for the hardwork.

**Pull request merged**